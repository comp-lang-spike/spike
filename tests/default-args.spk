#!/usr/bin/env spike

hgg1(required, answer = 42, foo = "hey", weird = answer) {
    printf("required is %d; ", required);
    printf("answer is %d; ", answer);
    printf("foo is %s; ", foo);
    printf("weird is %d.\n", weird);
}

hgg2(required, answer = 42, foo = "hey", weird = answer, ...extra) {
    printf("required is %d; ", required);
    printf("answer is %d; ", answer);
    printf("foo is %s; ", foo);
    printf("weird is %d", weird);
    if (extra.size > 0) {
        printf("; extra contains");
        extra do: [ :x | printf(" %d", x); ];
        printf(".\n");
    } else {
        printf(".\n");
    }
}

class Foo {
    hgg3(required, answer = self.defaultAnswer, foo = self.defaultFoo, weird = answer) {
        printf("required is %d; ", required);
        printf("answer is %d; ", answer);
        printf("foo is %s; ", foo);
        printf("weird is %d.\n", weird);
    }
    defaultAnswer { return 73; }
    defaultFoo { return "later"; }
}

f(x) { printf("[f(%d)] ", x); return x + 1; }

flora(x = 42, y = f(x)) {
    printf("whine: %d, %d\n", x, y);
}

check(hgg) {
    hgg(0);
    hgg(1, 24);
    hgg(2, 666, "you");
    hgg(3, 666, "guys", 999);
    if (hgg == hgg2) {
        hgg(4, 666, "you", 999, 77);
        hgg(5, 666, "guys", 999, 77, 88);
    }
}

hgg3_wrapper(...args) {
    obj f = Foo.new();
    f.hgg3(...args);
}

main(argv) {
    check(hgg1);
    check(hgg2);
    flora();
    flora(0);
    flora(2, 3);
    check(hgg3_wrapper);
    return 0;
}
