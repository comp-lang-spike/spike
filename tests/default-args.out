required is 0; answer is 42; foo is hey; weird is 42.
required is 1; answer is 24; foo is hey; weird is 24.
required is 2; answer is 666; foo is you; weird is 666.
required is 3; answer is 666; foo is guys; weird is 999.
required is 0; answer is 42; foo is hey; weird is 42.
required is 1; answer is 24; foo is hey; weird is 24.
required is 2; answer is 666; foo is you; weird is 666.
required is 3; answer is 666; foo is guys; weird is 999.
required is 4; answer is 666; foo is you; weird is 999; extra contains 77.
required is 5; answer is 666; foo is guys; weird is 999; extra contains 77 88.
[f(42)] whine: 42, 43
[f(0)] whine: 0, 1
whine: 2, 3
required is 0; answer is 73; foo is later; weird is 73.
required is 1; answer is 24; foo is later; weird is 24.
required is 2; answer is 666; foo is you; weird is 666.
required is 3; answer is 666; foo is guys; weird is 999.
